File: src/main/scala/akkastudy/supervisorstrategy/scala/AllForOneGrandparentActor.scala
package akkastudy.supervisorstrategy.scala

import akka.actor.{AllForOneStrategy, Actor, Props}

class AllForOneGrandparentActor extends Actor {

  import akka.actor.SupervisorStrategy._
  import scala.concurrent.duration._

  override val supervisorStrategy =
    AllForOneStrategy(maxNrOfRetries = 10, withinTimeRange = 1 minute) {
      case _: ArithmeticException ⇒ Resume
      case _: NullPointerException ⇒ Restart
      case _: IllegalArgumentException ⇒ Stop
      case _: Exception ⇒ Escalate
    }

  def receive = {
    case p: Props ⇒ sender ! context.actorOf(p)
  }
}

File: src/main/java/akkastudy/supervisorstrategy/java/AllForOneGrandparentActor.java
package akkastudy.supervisorstrategy.java;

import akka.actor.AllForOneStrategy;
import akka.actor.Props;
import akka.actor.SupervisorStrategy;
import akka.actor.UntypedActor;
import akka.japi.Function;
import scala.concurrent.duration.Duration;

import java.util.concurrent.TimeUnit;

public class AllForOneGrandparentActor extends UntypedActor {

    @Override
    public SupervisorStrategy supervisorStrategy() {
        return new AllForOneStrategy(10, Duration.create(1, TimeUnit.HOURS),
                new Function<Throwable, SupervisorStrategy.Directive>() {
                    @Override
                    public SupervisorStrategy.Directive apply(Throwable param) throws Exception {
                        if (param instanceof IllegalArgumentException) return SupervisorStrategy.stop();
                        if (param instanceof ArithmeticException) return SupervisorStrategy.resume();
                        if (param instanceof NullPointerException) return SupervisorStrategy.restart();
                        else return SupervisorStrategy.escalate();
                    }
                }
        );
    }

    @Override
    public void onReceive(Object message) throws Exception {
        if (message instanceof Props) {
            Props props = (Props) message;
            getSender().tell(getContext().actorOf(props), self());
        } else unhandled(message);
    }
}
